### ISSUE TITLE
Can't evaluate and continue training a DQN

### ISSUE THREAD

**Issue by Thrasher091** [deeplearning4j/deeplearning4j#7028]
#### Issue Description
Hi there! 

I have build a simple DQN in which an Agent collects blue circles. His environment is no gridworld, so I implemented a deep neural net (3 fully connected layers) with dl4j. 

For every circle he collects he gets +1 reward. 
After training the Agent for a while he gets results which are not as bad. He collects 160/200 in 30 secs.

After the training I save the model. Then I rewrite the code a little bit, so that he only preprocesses the states and decides what action to take, so that I can watch him play.

And thats where the problem occurs. After loading the Network he seems to have forgotten the learned policy. The Qvalues he give are nearly the same for all states and so he just take one action(which is the one with the highest Qvalue) again and again. 

Also when I don't rewrite the code and just want to continue training, the rewards drop from 160 to 5 because he just goes in one direction. Even when I start with a new Network and set Exploration to 0.1 he gets a score of 70, so why has suddenly one Qvalue such a high value after saving and loading the model?

The Iteration counter goes on from that moment I saved the Network ( like now 1895685) so the saving works...

My experienceMemory from which I train the Network is a list and I don't save her, because I think he should just collect new experiences when he starts to train again.

The Network:
` 

public class NeuralNetwork {
	

	int inputs=Testfeld.fields*Testfeld.parameter;
	int outputs =Testfeld.actionSize;
	
	double learningRate=Testfeld.learningRate;
	
	int[] hiddenLayers =new int[]{54,36,18};
	
	private MultiLayerNetwork MLN;
	
	public NeuralNetwork(){};

	
	
	public void createNetwork(){
			MultiLayerConfiguration configuration = null;
			NeuralNetConfiguration.ListBuilder builder = new NeuralNetConfiguration.Builder()
				
					
				
                .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
                .biasInit(0)
                .updater(new RmsProp.Builder().learningRate(learningRate).build())
              .biasUpdater(new RmsProp.Builder().learningRate(learningRate).build())
         
                .list()

                .layer(0, new DenseLayer.Builder()
                		.nIn(inputs)
                		.nOut(hiddenLayers[0])
                		.name("Input layer")
                        .weightInit(WeightInit.XAVIER)
                        .activation(Activation.RELU)
                        .name("Input Layer")
                        .build());
		
				builder.layer(1, new DenseLayer.Builder()
						.nIn(hiddenLayers[0])
						.nOut(hiddenLayers[1])
						.weightInit(WeightInit.XAVIER)
						.activation(Activation.RELU)
						.name("First hidden layer")
						.build());
				builder.layer(2, new DenseLayer.Builder()
						.nIn(hiddenLayers[1])
						.nOut(hiddenLayers[2])
						.weightInit(WeightInit.XAVIER)
						.activation(Activation.RELU)
						.name("Second hidden layer")
						.build());
				
				configuration =  builder.layer(3, new OutputLayer.Builder(LossFunction.MSE)
						.nIn(hiddenLayers[2])
						.nOut(outputs)
						.weightInit(WeightInit.XAVIER)
						.activation(Activation.IDENTITY)
						.weightInit(WeightInit.XAVIER)
						.name("Output layer")
						.build())
						
							
				.pretrain(false)
				.backprop(true)
				.build();

	MultiLayerNetwork network = new MultiLayerNetwork(configuration);
	
	network.init();
	

 network.setListeners(new ScoreIterationListener(Constants.NEURAL_NET_ITERATION_LISTENER));
   network.setListeners(new PerformanceListener(1));

	this.MLN = network;
	}

	

	public void trainNetwork(org.nd4j.linalg.dataset.api.iterator.DataSetIterator datasetIterator) {
		Nd4j.getMemoryManager().setAutoGcWindow(10000); 
		MLN.getLayerWiseConfigurations().setInferenceWorkspaceMode(WorkspaceMode.ENABLED);

		//MLN.setListeners(new PerformanceListener(1, true));
		MLN.fit(datasetIterator);
	}
	
	
	public void saveModel(String path) throws IOException {

	    File locationToSave = new File(path);      

	    boolean saveUpdater = true;                                  

	    ModelSerializer.writeModel(this.MLN, locationToSave, saveUpdater);

	}
	
	
	

	public void loadModel(String fileName) throws IOException {

		MultiLayerNetwork network = ModelSerializer.restoreMultiLayerNetwork(fileName, true);
		 network.setListeners(new ScoreIterationListener(Constants.NEURAL_NET_ITERATION_LISTENER));
		this.MLN=network;
    }
	
	
	
	
	public synchronized INDArray predictAction(INDArray input) {               

        INDArray resultVector = MLN.output(input);       

		return resultVector;

	}

}
`

DeepQNetwork:
`
public class DeepQNetwork  {
	
	private Random randomfunc = new Random();

	private NeuralNetwork cnn1;

	private NeuralNetwork cnn2;
	

	public DeepQNetwork(){

		this.cnn1= new NeuralNetwork();

		this.cnn2 = new NeuralNetwork();

	}

	public void generateNNs(){

		cnn1.createNetwork();
	
		cnn2=cnn1;
	}

	public void loadNNs(String filename) throws IOException{

		cnn1.loadModel(filename+"_q1");

		cnn2.loadModel( filename+"_q2");

	}

	public void saveNN(String filename) throws IOException{

		cnn1.saveModel(filename+"_q1");

		cnn2.saveModel(filename+"_q2");

	}

	public int selectAction(INDArray currentState, double explorationRate) {		

		int action=0;

		INDArray qValues1 = cnn1.predictAction(State.getFlattenState(currentState)); 
		INDArray qValues2 = cnn2.predictAction(State.getFlattenState(currentState));

    	INDArray qValues = qValues1.add(qValues2);   	

		double random = randomfunc.nextDouble();

		if (random < explorationRate) {

			action = randomfunc.nextInt(Testfeld.actionSize);	

		} else {			

			action=findMaxIndex(qValues);								

		}
		return action;

	}

	private int findMaxIndex(INDArray inputqValues){

		int returnq=0;
	
		double max = (inputqValues.maxNumber()).doubleValue();
	
		
			for(int i=0;i<Testfeld.actionSize;i++){

			double qvalue= inputqValues.getDouble(i);

			if(max==qvalue){

				returnq=i;

				break;

			}
		

		}

		return returnq;

	}

	public void learnOnMemory(Memory memory) {

		List<AddExperience> miniBatch = memory.getMiniBatch();

		List<Pair<INDArray, INDArray>> pairedList= new ArrayList<Pair<INDArray, INDArray>>();

	
		for (AddExperience entry : miniBatch) {

			INDArray qValues = Nd4j.zeros(Testfeld.actionSize,1);			

			INDArray qValuesObserved = Nd4j.zeros(Testfeld.actionSize,1);

			INDArray lastState= State.getFlattenState(entry.getState()); 
			
			INDArray observedState =State.getFlattenState(entry.getNewState());// same
			double rand = randomfunc.nextDouble();
			
			if (rand < 0.5) {

				qValues = cnn1.predictAction(lastState);
			
				qValuesObserved = cnn2.predictAction(observedState);
						} else {

				qValues = cnn2.predictAction(lastState);
			
				qValuesObserved = cnn1.predictAction(observedState);
						}

			double targetValue = calculateOptimalQ(qValuesObserved, entry.getReward(), entry.isTerminal());

		
			double oldvalue = qValues.getDouble(0,entry.getAction());
			

			
			double newValue=(1-Testfeld.learningRate)*oldvalue+ Testfeld.learningRate*targetValue;
			
		
			
			qValues.putScalar(0,entry.getAction(), newValue);
			
	
			
			Pair<INDArray,INDArray> pairedInputOutput= new Pair<INDArray, INDArray>(lastState, qValues);
			
		
			pairedList.add(pairedInputOutput);

		}
		
		DataSetIterator datasetIterator = new INDArrayDataSetIterator(pairedList, Testfeld.miniBatchSize);

		   

			cnn1.trainNetwork(datasetIterator);

	
	}

	private double calculateOptimalQ(INDArray qValuesObservedState, double reward, boolean isDone) {		

		
		
		double maxQ= (qValuesObservedState.maxNumber()).doubleValue();

        if (isDone)

            return reward;

        else

            return reward + Testfeld.gamma * maxQ;

}
    
 
    
    
}`

In the main file I start the Network with:
`
void DeepQLearnerSetup() throws IOException {
	//Create memory
	
	

	memory = new Memory();

	dqn = new DeepQNetwork();

	//Check whether the NNets are saved before

	File f1 = new File("lonelychase_q1");

	File f2 = new File("lonelychase_q2");		

	if(f1.exists()&&f2.exists()) {
	
		dqn.loadNNs("lonelychase");
	}
	else {

		dqn.generateNNs();
	}
}
`

and save it after all episodes with:

`	try {
			dqn.saveNN("lonelychase");
		} catch (IOException e1) {
			// TODO Auto-generated catch block
			e1.printStackTrace();
		} 
`

I really appreciate every help, because I haven't got much time left for this Project. 

Changing the files to .zip didn't help.

#### Version Information

* Deeplearning4j 1.0.0-beta3
* Windows 10

**Comment by Thrasher091** [comment]
To break down the problem a Little bit, if I train  the Agent with a new initilized Network only one game with almost no exploration, he gets a score from 70/200. 

When I than save and load the model, the Agent gets a score from 5/200 because there is one Qvalue which is the highest every time.

**Comment by saudet** [comment]
Duplicate of #7005
